reset-piv/reset-piv.c

```diff
@@ -0,0 +1,306 @@
+ */
+
+#include <bunyan.h>
+#include <err.h>
+#include <errno.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <unistd.h>
+#include "pivy/errf.h"
+#include "pivy/piv.h"
+
+void alloc_init(void);
+
+SCARDCONTEXT ctx;
+extern __thread bunyan_logger_t *tlog;
+
+static void
+usage(void)
+{
+	fprintf(stderr,
+	    "Usage: %s [guid]\n"
+	    "  If multiple initialized PIV tokens are present on the system, \n"
+	    "  the guid must be supplied\n",
+	    getprogname());
+
+	exit(EXIT_FAILURE);
+}
+
+static errf_t *
+parse_guid(const char *str, uint8_t guid[GUID_LEN])
+{
+	const char *p = str;
+	size_t n = 0;
+	size_t shift = 4;
+
+	while (*p != '\0') {
+		const char c = *p++;
+
+		if (n == GUID_LEN) {
+			return (errf("LengthError", NULL,
+			    "'%s' is not a valid GUID", str));
+		}
+
+		switch (c) {
+		case '0': case '1': case '2': case '3': case '4':
+		case '5': case '6': case '7': case '8': case '9':
+			guid[n] |= (c - '0') << shift;
+			break;
+		case 'A': case 'B': case 'C': case 'D': case 'E': case 'F':
+			guid[n] |= (c - 'A' + 10) << shift;
+			break;
+		case 'a': case 'b': case 'c': case 'd': case 'e': case 'f':
+			guid[n] |= (c - 'a' + 10) << shift;
+			break;
+		case ' ': case ':': case '\t':
+			continue;
+		default:
+			return (errf("InvalidCharacter", NULL,
+			    "%c is not a valid hex digit", c));
+		}
+
+		if (shift == 4) {
+			shift = 0;
+		} else {
+			shift = 4;
+			n++;
+		}
+	}
+
+	return (ERRF_OK);
+}
+
+static errf_t *
+get_piv(const char *guidstr, struct piv_token **ptp)
+{
+	errf_t *ret = ERRF_OK;
+	uint8_t guid[GUID_LEN] = { 0 };
+
+	*ptp = NULL;
+
+	if (guidstr != NULL) {
+		ret = parse_guid(guidstr, guid);
+		if (ret != ERRF_OK) {
+			errfx(EXIT_FAILURE, ret, "failed to parse guid '%s'",
+			    guidstr);
+		}
+
+		ret = piv_find(ctx, guid, GUID_LEN, ptp);
+		if (ret != ERRF_OK) {
+			return (errf("NotFoundError", ret,
+			    "cannot find PIV token '%s'", guidstr));
+		}
+	}
+
+	struct piv_token *tokens = NULL;
+	struct piv_token *t = NULL;
+
+	ret = piv_enumerate(ctx, &tokens);
+	if (ret != ERRF_OK) {
+		errfx(EXIT_FAILURE, ret, "failed to enumrate PIV tokens");
+	}
+
+	for (t = tokens; t != NULL; t = piv_token_next(t)) {
+		const uint8_t *tguid = piv_token_guid(t);
+
+		if (tguid == NULL)
+			continue;
+
+		if (guidstr != NULL && memcmp(tguid, guid, GUID_LEN) != 0)
+			continue;
+
+		if (*ptp != NULL) {
+			return (errf("CardError", NULL,
+			    "multiple PIV tokens present; must provide guid to "
+			    "reset"));
+		}
+
+		*ptp = t;
+	}
+
+	/* We end up leaking the rest of tokens, but that's ok */
+
+	if (*ptp == NULL) {
+		return (errf("NotFoundError", NULL,
+		    "no PIV tokens were found"));
+	}
+
+	return (ERRF_OK);
+}
+
+static char *
+piv_token_shortid(struct piv_token *pk)
+{
+	char *guid = NULL;
+
+	if (piv_token_has_chuid(pk)) {
+		guid = strdup(piv_token_guid_hex(pk));
+	} else {
+		guid = strdup("0000000000");
+	}
+	guid[8] = '\0';
```

Comment:
Check length with `strdup()` before assigning.  This can happen in the first leg of the `if` statement, presuming the zero guid at line 154 is shortened to 8 characters.

Maybe just write the value into a static buffer that is returned to avoid the leak.  Not that having a leak few cycles before the process dies is that big of a deal.

Reply:
I had copied it from pivy, but using a static buffer here would be simpler. I'll update accordingly.

Reply:
confirmed